File: apps/web/src/lib/audio/uploader.ts
/**
 * File d'envoi ordonnée : segments audio et événements (pause, reprise, trou).
 *
 * - un seul envoi à la fois, dans l'ordre de capture ;
 * - un élément ne quitte la file qu'après accusé de réception du serveur ;
 * - erreur temporaire (réseau, 5xx) : nouvelle tentative avec attente croissante,
 *   sans limite tant que la session vit (spec §69 : rien n'est abandonné en silence) ;
 * - erreur définitive : l'élément est déclaré perdu et signalé.
 */

import type { PcmChunk } from "./chunker";

export type GapReason = "microphone_lost" | "page_reloaded" | "capture_error";

export type QueueItem =
  | { kind: "chunk"; chunk: PcmChunk }
  | { kind: "pause" }
  | { kind: "resume" }
  | { kind: "gap"; reason: GapReason; durationMs: number | null };

export type SendResult = { ok: true } | { ok: false; retry: boolean; code: string };

export interface UploadTransport {
  send(item: QueueItem): Promise<SendResult>;
}

export interface UploaderStatus {
  pending: number;
  pendingBytes: number;
  retrying: boolean;
  lastErrorCode: string | null;
  lost: { item: QueueItem; code: string }[];
}

type Schedule = (callback: () => void, delayMs: number) => void;

export const DEFAULT_BACKOFF_MS = [500, 1000, 2000, 4000, 8000];

export class Uploader {
  private queue: QueueItem[] = [];
  private running = false;
  private stopped = false;
  private retrying = false;
  private lastErrorCode: string | null = null;
  private lost: { item: QueueItem; code: string }[] = [];
  private listeners = new Set<(status: UploaderStatus) => void>();
  private drainWaiters: (() => void)[] = [];

  constructor(
    private readonly transport: UploadTransport,
    private readonly backoffMs: number[] = DEFAULT_BACKOFF_MS,
    private readonly schedule: Schedule = (callback, delay) => void setTimeout(callback, delay),
  ) {}

  enqueue(item: QueueItem): void {
    this.queue.push(item);
    this.notify();
    void this.run();
  }

  get status(): UploaderStatus {
    return {
      pending: this.queue.length,
      pendingBytes: this.queue.reduce(
        (total, item) => total + (item.kind === "chunk" ? item.chunk.samples.length * 2 : 0),
        0,
      ),
      retrying: this.retrying,
      lastErrorCode: this.lastErrorCode,
      lost: [...this.lost],
    };
  }

  subscribe(listener: (status: UploaderStatus) => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  /** Résout quand la file est vide. */
  whenDrained(): Promise<void> {
    if (this.queue.length === 0) return Promise.resolve();
    return new Promise((resolve) => this.drainWaiters.push(resolve));
  }

  /** Abandon explicite (le praticien termine malgré des envois impossibles). */
  abandon(): QueueItem[] {
    const abandoned = this.queue;
    this.queue = [];
    this.stopped = true;
    this.retrying = false;
    this.notify();
    this.resolveDrain();
    return abandoned;
  }

  private async run(): Promise<void> {
    if (this.running || this.stopped) return;
    this.running = true;
    let attempt = 0;
    while (this.queue.length > 0 && !this.stopped) {
      const item = this.queue[0]!;
      let result: SendResult;
      try {
        result = await this.transport.send(item);
      } catch {
        result = { ok: false, retry: true, code: "NETWORK_UNREACHABLE" };
      }
      if (this.stopped) break;
      if (result.ok) {
        this.queue.shift();
        attempt = 0;
        this.retrying = false;
        this.lastErrorCode = null;
      } else if (result.retry) {
        this.retrying = true;
        this.lastErrorCode = result.code;
        this.notify();
        const delay = this.backoffMs[Math.min(attempt, this.backoffMs.length - 1)] ?? 1000;
        attempt += 1;
        await new Promise<void>((resolve) => this.schedule(resolve, delay));
        continue;
      } else {
        this.queue.shift();
        this.lost.push({ item, code: result.code });
        this.lastErrorCode = result.code;
        attempt = 0;
      }
      this.notify();
    }
    this.running = false;
    if (this.queue.length === 0) this.resolveDrain();
  }

  private resolveDrain(): void {
    const waiters = this.drainWaiters;
    this.drainWaiters = [];
    waiters.forEach((resolve) => resolve());
  }

  private notify(): void {
    const status = this.status;
    this.listeners.forEach((listener) => listener(status));
  }
}

File: apps/ios/Oris/Audio/Uploader.swift
import Foundation

struct UploadStatus: Sendable, Equatable {
    var pending: Int
    var retrying: Bool
    var lastErrorCode: String?
    var lost: Int
}

/// File d'envoi ordonnée : un seul envoi à la fois, dans l'ordre de capture.
///
/// Un segment est d'abord écrit dans le tampon chiffré, puis envoyé ; il n'est
/// supprimé qu'après accusé de réception. Erreur temporaire : nouvel essai avec
/// attente croissante, sans limite tant que la session vit (spec §69). Erreur
/// définitive : segment déclaré perdu (le serveur signalera le trou).
actor Uploader {
    private let encounterId: String
    private let store: ChunkStore
    private let transport: any AudioTransporting
    private let backoff: [Duration]
    private let onStatus: @Sendable (UploadStatus) -> Void

    private var queue: [UploadItem] = []
    private var running = false
    private var stopped = false
    private var retrying = false
    private var lastErrorCode: String?
    private var lost = 0
    private var drainWaiters: [CheckedContinuation<Void, Never>] = []
    private var sleepTask: Task<Void, Never>?

    init(
        encounterId: String,
        store: ChunkStore,
        transport: any AudioTransporting,
        backoff: [Duration] = [.milliseconds(500), .seconds(1), .seconds(2), .seconds(4), .seconds(8)],
        onStatus: @escaping @Sendable (UploadStatus) -> Void = { _ in }
    ) {
        self.encounterId = encounterId
        self.store = store
        self.transport = transport
        self.backoff = backoff
        self.onStatus = onStatus
    }

    var status: UploadStatus {
        UploadStatus(pending: queue.count, retrying: retrying, lastErrorCode: lastErrorCode, lost: lost)
    }

    /// Enregistre le segment chiffré puis le met en file.
    func enqueue(chunk: PcmChunk) throws {
        try store.save(chunk, encounterId: encounterId)
        append(.chunk(sequence: chunk.sequence))
    }

    func enqueue(_ item: UploadItem) {
        append(item)
    }

    /// Réouverture de l'app : les segments restés dans le tampon repartent en tête de file.
    func restorePending() -> [Int] {
        let queued = Set(queue.compactMap { item -> Int? in
            if case .chunk(let sequence) = item { return sequence }
            return nil
        })
        let restored = store.pendingSequences(encounterId: encounterId).filter { !queued.contains($0) }
        queue.insert(contentsOf: restored.map { UploadItem.chunk(sequence: $0) }, at: 0)
        publish()
        start()
        return restored
    }

    /// Le réseau est revenu : on n'attend pas la fin du délai en cours.
    func wake() {
        sleepTask?.cancel()
    }

    func waitUntilDrained() async {
        if queue.isEmpty { return }
        await withCheckedContinuation { drainWaiters.append($0) }
    }

    /// « Terminer malgré tout » : ce qui n'est pas parti est abandonné et effacé localement.
    func abandon() {
        for item in queue {
            if case .chunk(let sequence) = item {
                store.delete(encounterId: encounterId, sequence: sequence)
            }
        }
        queue.removeAll()
        stopped = true
        retrying = false
        sleepTask?.cancel()
        publish()
        resolveDrain()
    }

    #if DEBUG
    /// Tests : simule une app tuée (file perdue, tampon chiffré conservé).
    func abandonWithoutDeletingForTest() {
        queue.removeAll()
        stopped = true
        sleepTask?.cancel()
    }
    #endif

    private func append(_ item: UploadItem) {
        guard !stopped else { return }
        queue.append(item)
        publish()
        start()
    }

    private func start() {
        guard !running, !stopped else { return }
        running = true
        Task { await self.run() }
    }

    private func run() async {
        var attempt = 0
        while let item = queue.first, !stopped {
            var chunk: PcmChunk?
            var result: SendResult
            if case .chunk(let sequence) = item {
                chunk = try? store.load(encounterId: encounterId, sequence: sequence)
            }
            if case .chunk = item, chunk == nil {
                result = .failed(code: "LOCAL_CHUNK_UNREADABLE", retry: false)
            } else {
                result = await transport.send(item, chunk: chunk)
            }
            if stopped { break }
            switch result {
            case .ok:
                queue.removeFirst()
                if case .chunk(let sequence) = item {
                    store.delete(encounterId: encounterId, sequence: sequence)
                }
                attempt = 0
                retrying = false
                lastErrorCode = nil
            case .failed(let code, retry: true):
                retrying = true
                lastErrorCode = code
                publish()
                let delay = backoff[min(attempt, backoff.count - 1)]
                attempt += 1
                let task = Task { _ = try? await Task.sleep(for: delay) }
                sleepTask = task
                await task.value
                continue
            case .failed(let code, retry: false):
                queue.removeFirst()
                if case .chunk(let sequence) = item {
                    store.delete(encounterId: encounterId, sequence: sequence)
                }
                lost += 1
                lastErrorCode = code
                attempt = 0
            }
            publish()
        }
        running = false
        if queue.isEmpty { resolveDrain() }
    }

    private func resolveDrain() {
        let waiters = drainWaiters
        drainWaiters = []
        waiters.forEach { $0.resume() }
    }

    private func publish() {
        onStatus(status)
    }
}
